### ISSUE TITLE
Authentication

### ISSUE THREAD

**Issue by dotNETSanta** [CoyoteNET/CoyoteNET-backend#17]
This issue tries to analyze everything that's related to: our **Authentication** layer

So far here's previous discussion:

>Additionally, I'm not a huge fan of standard Microsoft identity implementation for login/register/etc. We can do our authentication.

_Originally posted by @Kubion70 in https://github.com/CoyoteNET/CoyoteNET-backend/issues/2#issuecomment-476824324_

>You mean jwt and writing our own auth layer (pw hashing, locking after a few failed login attempts, recovery tokens, so on)?

_Originally posted by @dotNETSanta in https://github.com/CoyoteNET/CoyoteNET-backend/issues/2#issuecomment-476826506_

>Yes. I've done it several times so I can handle it if needed.

_Originally posted by @Kubion70 in https://github.com/CoyoteNET/CoyoteNET-backend/issues/2#issuecomment-476826797_

>Authentication : on the beginning simple base64 header with login and password should work great. I can't imagine that someone will use it with "real" password. On the next versions, we can easily replace it with JWT, oauth etc.

_Originally posted by @orchowski in https://github.com/CoyoteNET/CoyoteNET-backend/issues/2#issuecomment-476860779_

**Comment by smknd** [comment]
Wouldn't be easier to start with GitHub account integration and after that think about custom mechanism?

**Comment by silvuss** [comment]
@smknd What did you exactly mean?

**Comment by mateuszradny** [comment]
Probably some people will not want to combine their own github account with the forum, so I suggest adding a Github account login option as an additional feature.

**Comment by bednarz1990** [comment]
Nope, GitHub login as default auth is silly idea at this stage. Github auth should be additional option for users.

**Comment by xhexe** [comment]
The default login methods are always weak. However the ability to login with external services should exist. For example:

- Github authentication should use only login and e-mail

Using the external authentication should not create an account. User can should decide if he wants to improve some data.
And of course the first option to register should be using e-mail and password. The remaining fields can be updated after the first login.

Additional functionality is the ability to disconnect or attach the external account in the user profile - * low priority *.

**Comment by smknd** [comment]
At CURRENT stage, everyone involved in the project has a GitHub account. That's why I suggested to use this for now, because everyone can easily use it for testing purposes without wasting the time on reinventing the wheel.

**Comment by bednarz1990** [comment]
Yes, @smknd you are right, but I think we should focus for future users and don't waste time for writing two auth methods. Let's stay in asp.net auth so far 💯

**Comment by ProximaB** [comment]
The thing is that we're trying to engage as many people as possible at once. There are some milestones that need to well-planned and implement immediately e.g database.  We should create some roadmap of projects features and transform that to tasks. @smknd may be right, if it was possible to implement OAuth sooner and take on other tasks then it's worth doing, in the meantime, someone may develop another type of authentication.

**Comment by silvuss** [comment]
@ProximaB There is already a roadmap on the wiki: https://github.com/CoyoteNET/CoyoteNET-backend/wiki/CoyoteNET-roadmap, may you make a judgment. :)

**Comment by smknd** [comment]
@bednarz1990: The primary reason of my suggestion is that I don't want to waste time. AFAIK ASP.NET already has support for external OAuth providers like GitHub.
